### ISSUE TITLE
Large file uploads fail with `ClientException (Connection closed before full header was received)`

### ISSUE THREAD

**Issue by lukehutch** [serverpod/serverpod#981]
**Describe the bug**

When I try to upload a file to the Serverpod server, the server successfully logs

```
METHOD CALL: uploadFile.verifyUpload duration: 9ms numQueries: 0 authenticatedUser: 1
```

but then the client fails at `FileUploader.upload`, where the `http.post` method throws:

```
ClientException (Connection closed before full header was received)
```

**Serverpod versions**
 -  _CLI Version:_ 1.0.1
 - _Library version:_ 1.0.1

**Platform information**

Android emulator / Linux

**Comment by lukehutch** [comment]
This happens on every file upload.

Not sure if you are OK with adding dependencies to Serverpod, but this guy had the same problem, and solved it by switching to Dio: https://stackoverflow.com/questions/54844472/http-connection-issue

**Comment by lukehutch** [comment]
This was really hard to debug, but I traced this issue down to the following:

In `public_endpoint.dart`, class `CloudStoragePublicEndpoint`:

```dart
  Future<List<int>?> _readBinaryBody(HttpRequest request) async {
    // TODO: Find more efficient solution?
    var len = 0;
    var data = <int>[];

    await for (var segment in request) {
      len += segment.length;
      if (len > server.serverpod.config.maxRequestSize) return null;   // (*)
      data += segment;
    }
    return data;
  }
```

See the line `// (*)`.

The upload fails when one of my upload segments has size `557663`, and `server.serverpod.config.maxRequestSize` has value `524288`.

Why is this segment size limit enforced? Shouldn't this be handled more gracefully?

**Comment by pro100andrey** [comment]
To increase the maximum request size, set the value of maxRequestSize in the config/development.yaml file. Example:

``` yaml
#config/development.yaml 

maxRequestSize: 52428800

# Configuration for the main API server.
apiServer:
  port: 8080
  publicHost: localhost
  publicPort: 8080
  publicScheme: http
```

**Comment by lukehutch** [comment]
@pro100andrey Thank you! I'll close this issue, since the remaining ones are covered by your PRs plus my comments on them.

**Comment by lukehutch** [comment]
The other bug about this was closed for some reason, so I'll reopen this one..

**Comment by SandPod** [comment]
> At the very least, request size exceeded needs to be logged on the server, and the request should return a 413 Content Too Large result code, rather than the server just returning false.

@lukehutch since the file size is configurable, documentation [here](https://docs.serverpod.dev/next/concepts/working-with-endpoints#passing-parameters) and [here](https://docs.serverpod.dev/next/concepts/configuration#configuration-options), would you agree that the quote captures the task of this issue?

**Comment by lukehutch** [comment]
@SandPod No, it's still not sufficient just to document this parameter. It was a pain in the butt to try to track down why uploads were failing. See this point above:

> At the very least, `request size exceeded` needs to be logged on the server, and the request should return a `413 Content Too Large` result code, rather than the server just returning `false`.

**Comment by SandPod** [comment]
I'm not sure I follow here. It seems like we have the same quote in our comments, could you clarify what more would be required than what is captured in the quote we both reference?

**Comment by lukehutch** [comment]
Oops, sorry, I was distracted when I responded... Yes, you identified the core issue.

**Comment by binemmanuel** [comment]
I experienced this issue today and figured out it was the file size from quickly cropping and trying the upload again. I believe if exceptions are raised for cases like this, it'll save a lot of development time, as just returning bool is not sufficient.
